### PR TITLE
[CBRD-24182] Remove unnecessary table such as outer joined table with unique predicates

### PR DESCRIPTION
**Pull request by shparkcubrid**
http://jira.cubrid.org/browse/CBRD-24182

The case of outer joined table with unique predicates.

```
SELECT count(*)  <== referenced column from tbl2 is not exists.
   FROM  tbl1 left outer join tbl2  <== tbl2 is outer joined.
 on tbl1.col1 = tbl2.PK <== tbl2.PK is unique predicate.
```
In this case, tbl2 can be removed.

How to check unique predicates.
 - All columns of a unique constraint must exist in the on_cond clause.
 - Only predcates with the following conditions are included.
 |              arg1                       |       op      |          arg2                      |
 | only have one of my spec  |       '='      |  not have my spec node|
 | e.g.)      col(my spec)          |        =       |  2 (constant)                  |
 |            a.col(my spec)         |        =       |  b.col                             |
 |            a.col(my spec)         |        =       |  b.col + c.col                 |

Restrictions
 - CTE spec
 - Right outer join

### CODE REVIEW

File: src/optimizer/query_rewrite.c [hunk 1]
```diff
@@ -3074,6 +3088,410 @@ qo_reduce_comp_pair_terms (PARSER_CONTEXT * parser, PT_NODE ** wherep)
...
+ *
+ * Note:
+ */
+static void
+qo_reset_spec_location (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE * query)
+{
+  short curr_loc, after_loc;
+  PT_NODE *where;
+  RESET_LOCATION_INFO locate_info;
+
```
Comment (byungwook-kim) [inline src/optimizer/query_rewrite.c:3291]: PT_NODE *p;

Comment (beyondykk9) [inline src/optimizer/query_rewrite.c:3291]:
```suggestion
PT_NODE *sp

sp = spec;
while (sp != NULL)
    {
      ...
      sp = sp->next;
    }
```

Comment (shparkcubrid) [reply src/optimizer/query_rewrite.c:3291]: Since the spec is only used here, I think it would be ok to keep it.

**Review by beyondykk9 — APPROVED** [review]

**Review by hyunikn — APPROVED** [review]

**Review by ctshim — APPROVED** [review]

File: src/parser/parse_tree_cl.c [hunk 2]
```diff
@@ -3103,6 +3103,27 @@ parser_append_node (PT_NODE * node, PT_NODE * list)
   return list;
 }
 
+/*
+ * parser_append_previous_node() -
+ *   return:
+ *   node(in/out):
```
Comment (hgryoo) [inline src/parser/parse_tree_cl.c]:
```suggestion
 *   node(in):
```

File: src/parser/parse_tree_cl.c [hunk 3]
```diff
@@ -3103,6 +3103,27 @@ parser_append_node (PT_NODE * node, PT_NODE * list)
...
+  if (list)
+    {
+      node->next = list;
+      list = node;
+    }
+  else
+    {
+      list = node;
+    }
+  return list;
```
Comment (hgryoo) [inline src/parser/parse_tree_cl.c:3124]: It is shortened version.
```suggestion
  if (list)
    {
      node->next = list;
      node = list;
    }
  return node;
```

File: src/optimizer/query_rewrite.c [hunk 4]
```diff
@@ -3074,6 +3088,410 @@ qo_reduce_comp_pair_terms (PARSER_CONTEXT * parser, PT_NODE ** wherep)
...
+ *                select count(*)
+ *                  from tbl1 a
+ */
+static PT_NODE *
+qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE * query)
+{
+  MOP cls;
+  SM_CLASS_CONSTRAINT *consp;
+  SM_ATTRIBUTE *attrp;
+  PT_NODE *point_list, *point, *where, *col, *tmp_spec, *prev_spec, *pred, *prev_pred, *next_pred;
```
Comment (hgryoo) [inline src/optimizer/query_rewrite.c]:
```suggestion
  PT_NODE *point_list = NULL, *point, *where, *col, *tmp_spec, *prev_spec, *pred, *prev_pred, *next_pred;
```

at the `end` flag, point_list is freed according to the case this variable is NULL.

**Review by hgryoo — APPROVED** [review]

File: src/optimizer/query_rewrite.c (hunk 4 above)
Comment (shparkcubrid) [reply src/optimizer/query_rewrite.c]: I'll fix it.

File: src/parser/parse_tree_cl.c (hunk 2 above)
Comment (shparkcubrid) [reply src/parser/parse_tree_cl.c]: I'll fix it.

File: src/parser/parse_tree_cl.c (hunk 3 above)
Comment (shparkcubrid) [reply src/parser/parse_tree_cl.c:3124]: I think it is not same.

**Review by youngjinj — APPROVED** [review]

File: src/parser/parse_tree_cl.c (hunk 3 above)
Comment (hgryoo) [reply src/parser/parse_tree_cl.c:3124]: Assigning to the `list` variable changes the local variable, so there is no change outside this function.

Comment (hgryoo) [reply src/parser/parse_tree_cl.c:3124]: I wonder if you are intending to change the reference of the list variable passed when calling this function.

Comment (shparkcubrid) [reply src/parser/parse_tree_cl.c:3124]: `list` is return var. so I think it is okay.

**Pull request merged**